etcdserver/security/security.go

```diff
@@ -0,0 +1,303 @@
+	Permissions Permissions `json:"permissions"`
+	Grant       Permissions `json:"grant"`
+	Revoke      Permissions `json:"revoke"`
+	Modified    time.Time   `json:"lastModified"`
+}
+
+type Permissions struct {
+	KV rwPermission `json:"kv"`
+}
+
+type rwPermission struct {
+	Read  []string `json:"read"`
+	Write []string `json:"write"`
+}
+
+func mergeErr(s string, v ...interface{}) *httptypes.HTTPError {
+	return httptypes.NewHTTPError(400, fmt.Sprintf(s, v...))
+}
+
+func NewStore(server *etcdserver.EtcdServer, timeout time.Duration) *Store {
+	s := &Store{
+		server:  server,
+		timeout: timeout,
+	}
+	s.ensureSecurityDirectories()
+	return s
+}
+
+func (s *Store) AllUsers() ([]string, error) {
+	resp, err := s.requestResource("/users/", false)
+	if err != nil {
+		return nil, err
+	}
+	var nodes []string
+	nodes = append(nodes, "root")
+	for _, n := range resp.Event.Node.Nodes {
+		_, user := path.Split(n.Key)
+		nodes = append(nodes, user)
+	}
+	sort.Strings(nodes)
+	return nodes, nil
+}
+
+func (s *Store) GetUser(name string) (User, error) {
+	// TODO(barakmich): Possibly add a cache
+	resp, err := s.requestResource("/users/"+name, false)
+	if err != nil {
+		return User{}, err
+	}
+	var u User
+	err = json.Unmarshal([]byte(*resp.Event.Node.Value), &u)
+	if err != nil {
+		return u, err
+	}
+
+	return u, nil
+}
+
+func (s *Store) CreateUser(user User) error {
+	hash, err := bcrypt.GenerateFromPassword([]byte(user.Password), bcrypt.DefaultCost)
+	if err != nil {
+		return err
+	}
+	user.Password = string(hash)
+	user.Modified = time.Now()
+
+	_, err = s.createResource("/users/"+user.User, user)
+	return err
+}
+
+func (s *Store) DeleteUser(name string) error {
+	_, err := s.deleteResource("/users/" + name)
+	return err
+}
+
+func (s *Store) UpdateUser(user User) (User, error) {
+	old, err := s.GetUser(user.User)
+	if err != nil {
+		return old, err
+	}
+	newUser, err := old.Merge(user)
+	if err != nil {
+		return old, err
+	}
+	_, err = s.updateResource("/users/"+user.User, newUser)
+	if err != nil {
+		return newUser, err
+	}
+	return newUser, nil
+}
+
+func (s *Store) AllRoles() ([]string, error) {
+	resp, err := s.requestResource("/roles/", false)
+	if err != nil {
+		return nil, err
+	}
+	var nodes []string
+	nodes = append(nodes, "root")
+	for _, n := range resp.Event.Node.Nodes {
+		_, role := path.Split(n.Key)
+		nodes = append(nodes, role)
+	}
+	sort.Strings(nodes)
+	return nodes, nil
+}
+
+func (s *Store) GetRole(name string) (Role, error) {
+	// TODO(barakmich): Possibly add a cache
+	resp, err := s.requestResource("/roles/"+name, false)
+	if err != nil {
+		return Role{}, err
+	}
+	var r Role
+	err = json.Unmarshal([]byte(*resp.Event.Node.Value), &r)
+	if err != nil {
+		return r, err
+	}
+
+	return r, nil
+}
+
+func (s *Store) CreateRole(role Role) error {
+	role.Modified = time.Now()
+	_, err := s.createResource("/roles/"+role.Role, role)
+	return err
+}
+
+func (s *Store) DeleteRole(name string) error {
```

Comment:
I've been thinking about that. Keeping that in sync is potentially a lot of writes. It is, however, correct.